### DISCUSSION TITLE
quarkus-keycloak-authorization will break quarkus-websockets. Why?

### DISCUSSION THREAD

**Discussion by SetoKaiba** [quarkusio/quarkus#30905] [Q&A] [answered]
When quarkus.keycloak.policy-enforcer.enable=true is set, the websocket with the path in keycloak authorization will not able to be connect.
The permission of the resource is the same with the websocket. Websocket doesn't support custom header. I think that webSocket doesn't support keycloak security. Why is it affacted? Or does the websocket support security?

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
keycloak-authorization requires a verified OIDC token to function

**Reply by SetoKaiba** [reply]
> keycloak-authorization requires a verified OIDC token to function

But I think the websocket does not support bearer token authorization because it doesn't support custom header. How does it verify the authorization for websocket?

**Reply by SetoKaiba** [reply]
I mean that if authorization is not supported. Why is websocket affacted?
I add a permission for path "/aaa". But it affacts both for rest "/aaa" and websocket "/aaa".
So the websocket will be unable to connect. Because no bearer token in the header since it doesn't support custom headers.
I thought that the websocket "/aaa" should not be affacted.

**Comment by pedroigor** [comment]
@sberyozkin @sberyozkin Perhaps the policy enforcer should only handle REST requests?

**Reply by sberyozkin** [reply]
Hey @pedroigor Right, I agree, but users would like to retain the token in the ws channel after the initial request and then have it handled as if it came with http... 
@SetoKaiba Custom approach is required, please check Discussions, as I believe a few threads are open.

**Comment by sberyozkin** [comment]
@SetoKaiba For example, #30427.
However it will be no-trivial to integrate it with the keycloak authorization which is implemented as HTTP sec policy check. Please try to make it work first with OIDC bearer token verification alone

**Reply by SetoKaiba** [reply]
Thank you for the info. 

![image](https://user-images.githubusercontent.com/159540/217120051-fa17aad7-1313-41c1-8088-84ab2fc102a7.png)
I use the message to achieve OIDC bearer token verification. Good to know that there's a way to use the subprotocol of WebSocket.
Which one is recommended? It seems that the subprotocol approach can take advantage of the keycloak authorization. Mine is just about the security identity.

And also, I'd like to add the authorization header whenever the subprotocol include the bearer token.
Is the @RouteFilter is just the same with the code below which doesn't need dependency with quarkus-reactive-routes?
```
void setupFilter(@Observes Filters filters) {
        filters.register((rc) -> {
           // ...

            rc.next();
        }, priority);
    }
```

**Reply by SetoKaiba** [reply]
@sberyozkin Hi. I try to achieve the subprotocol way. The modifyHandshake method is never in if the subprotocol is set in the client. Any idea? Thank you.
**EDIT:** Solved with the below post. But I'm still confused.

**Reply by SetoKaiba** [reply]
@sberyozkin It seems that the problem is the line below.
```
quarkus.keycloak.policy-enforcer.enable=true
```

How does rest authorization work with no Authorization set(no resource, no policy, no permission) in Keycloack?
How does websocket authorization work with no Authorization set(no resource, no policy, no permission) in Keycloak?

It won't let the rest unable to access.
But it will let the websocket unable to access.
I mean that it seems that they work different.
Rest allow the access without Authorization set.
WebSocket deny the access without Authorization set.
I think that they should be the same.

I'm sure that the authorization is worked(I got securityIdentity) without the policy-enforcer enabled.

**Reply by SetoKaiba** [reply]
@sberyozkin And also, how can I run a long task in WebSocket onOpen with SecurityIdentity?
If I don't make it a sub thread, it will costs more than 2 seconds. So it break the task.
If I make it a sub thread, it reports the error below.
```
jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of CLASS bean [class=io.quarkus.security.runtime.SecurityIdentityProxy, id=3fe7767cea7608ae785d5ac61e84fcc0f58d1622]
```

**Reply by sberyozkin** [reply]
>I'm sure that the authorization is worked(I got securityIdentity) without the policy-enforcer enabled.

Did you get it in the initial HTTP to WS upgrade request ? That is OK, HTTP is still live, or do you mean you got the token successfully carried around in the established WS channel ?

**Reply by SetoKaiba** [reply]
> > I'm sure that the authorization is worked(I got securityIdentity) without the policy-enforcer enabled.
> 
> Did you get it in the initial HTTP to WS upgrade request ? That is OK, HTTP is still live, or do you mean you got the token successfully carried around in the established WS channel ?

![image](https://user-images.githubusercontent.com/159540/217246332-b95a7e65-74c6-41fa-9c9c-0e3d2896339e.png)
I'm not sure. I think not. This is the WS line in developer tools.

If I disabled policy-enforcer. Then the autorization is work. The token is successfully carried around the established WS channel.

I'm confused that the policy enforcer works different for REST and WS.
Without any permission set, the REST is available to access.
Without any permission set, the WS is not available to access. If I add the /liveSocket/* resource, policy grant client, permission combine it. Then it will work.

I think that the WS should work like the REST, if permission is set, it check the permission. If not, it should just grant.

**Reply by sberyozkin** [reply]
I'm sorry, I'm not sure what I can help with. If you'd like to have WebSockets carry around the original bearer access token which was provided over HTTP during the upgrade request then please investigate the proposed solution in the discussion linked to earlier.

**Reply by SetoKaiba** [reply]
If I provide a reproduce project and realm settings, can you please have a look what's the problem?
**EDIT:** My mistake. Just check the below post.

**Reply by SetoKaiba** [reply]
It seems that I make a mistake. Both REST and WS will not grant when no permission is set.
Previously I just check it without Authorization header. My mistake.

Is there a way to make them all available without permission set with only the resource with permission checked?

**Reply by sberyozkin** [reply]
The problem is that you do not have WebSockets keeping the token, it is not really related to Keycloak at all.

**Reply by sberyozkin** [reply]
Please focus on WebSockets first, even without any security at all, make it work such that some custom property can be retained in the WS channel across WS requests. Once you have it working we can discuss how to handle the security token instead of the custom test property. So, please get the basic WS test working first and then we'll move on to its linking with the security system

**Reply by SetoKaiba** [reply]
No. Let me clarify it. I think it's just a policy enforcer mechanism problem
Because it works the same for REST and WS 
The Authorization header for WS is ok and I tested.

**It acts:**
- No resource, policy and permission.
  - With Authorizaion: Bearer xxx
     - REST(401) and WS(no upgrade) both not available.
  - Without Authorization: Bearer xxx
     - REST and WS both OK and I get anonymous SecurityIdentity
- Resource set for REST and WS. Policy grant client 'test-auth'. Permission combine them.
  -  With Authorization: Bearer xxx
     - REST and WS both OK and I can get SecurityIdentity
  - Without Authorization: Bearer xxx
     - REST(401) and WS(no upgrade) both not available.

**I expect:**
- No resource, policy and permission (or any settings that I can make it default avaiable for both)
  - With Authorizaion: Bearer xxx
    - REST and WS both OK and I can get SecurityIdentity
  - Without Authorization: Bearer xxx
    - REST and WS both OK and I get anonymous SecurityIdentity

**Comment by sberyozkin** [comment]
Would you mind linking to a WS-only application which confirms you have the custom property retained after the upgrade ?

**Reply by SetoKaiba** [reply]
I think the problem is not WS related now. But about REST also.

I previously use a client login message for WS.
When the client login is with Bearer token, I'll get the securityIdentity for it.
When the client login is without Bearer toekn. I'll just keep it anonymous.

This is what I expected for both REST and WS.

**Reply by SetoKaiba** [reply]
Please check the last post of your comment above. You can get what I actucally expect.

**Reply by sberyozkin** [reply]
I'd just like to make sure it is working before talking about combining WS and Keycloak authorization. As I said earlier, please make sure you have a concept working in principle, just retain `a=b` property in the web socket channel and verify it is available to the endpoint after the upgrade, on some regular WS channel call.

**Reply by SetoKaiba** [reply]
> I'd just like to make sure it is working before talking about combining WS and Keycloak authorization. As I said earlier, please make sure you have a concept working in principle, just retain `a=b` property in the web socket channel and verify it is available to the endpoint after the upgrade, on some regular WS channel call.

Like a previously said. Yes. It works. I can just disable the policy enforcer.
  - With Authorizaion: Bearer xxx
    - REST and WS both OK and I can get SecurityIdentity
  - Without Authorization: Bearer xxx
    - REST and WS both OK and I get anonymous SecurityIdentity

**Reply by SetoKaiba** [reply]
What I expect 
- All not listed resource in Keycloak Authoriztion tab grant with actual SecurityIdentity or anonymous SecurityIdentity.
- Only the resources set in Keycloak Authorization tab perform a permission check.

**Reply by sberyozkin** [reply]
Sure, let me have a look at the working WS-only test app first

**Reply by sberyozkin** [reply]
Also, as I said, keycloak authorization is implemented as HTTP Security policy check, so it can't work with non-HTTP calls, unless some specific integration is done.

**Reply by sberyozkin** [reply]
I can also suggest that you debug https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java#L44 just to see what is going on

**Reply by SetoKaiba** [reply]
https://github.com/SetoKaiba/reproduce
@sberyozkin  Hi. Here's the reproduce project. I use 8880 for reproduce project and 8180 for keycloak. I use fix port. So I run the keycloak docker and import the realm-export.json manually.
```
docker run -p 8180:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:20.0.3 start-dev --features=preview
```

http://localhost:8180/realms/test/account/#/
http://localhost:8880/
user: test
pass: test
1. Login, click 4 buttons and check developer tools.
2. Logout, click 4 buttons and check devloper tools.
3. Enable policy enforcer, login, click 4 buttons and check developer tools.
4. Enable policy enforcer, logout, click 4 buttons and check developer tools.

**Reply by SetoKaiba** [reply]
> I can also suggest that you debug https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java#L44 just to see what is going on

Thank you for your instruction.

I found strange thing. Enable policy enforcer.
I access to http://localhost:8880. The apply method is called.
I click the Request REST with Bearer. The apply method is not called. But it just return 401.

**Reply by SetoKaiba** [reply]
It seems that the realm I export is without user and the client secret is missed. I don't know how to export a full realm config. But you can set it yourself.

**Comment by sberyozkin** [comment, marked as answer]
@SetoKaiba Your WS code looks fine, thanks. Please confirm that you can also get the identity available in `onMessage()` which is critical. If that works, then it is great, you have WS and OIDC working. It will be definitely worth documenting, I proposed it to the user who provided a similar solution earlier, but I'll have np documenting it myself.

>I found strange thing. Enable policy enforcer....

Lets confirm first `onMessage()` can access a non-anonymous identity.
I honestly have no idea right now why KC enforcer is not seeing it. Or if https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java#L46 is even invoked - which also will have to be fixed (add log messages).

So confirm please `onMessage` can work with the non-anonymous identity, and add a debug breakpoint in the policy enforcer, it will be much clearer what is going on afterwards

**Reply by SetoKaiba** [reply]
Hi. I submitted a commit with onMessage test. It's working with non-anonymous identity. I mean with no policy enforcer.

**Reply by SetoKaiba** [reply]
![image](https://user-images.githubusercontent.com/159540/217297328-d83b8a6b-5c0f-4be2-bdc5-b3c3f07969be.png)

After a deeper dig, I think that here I should choose mode to permissive. But it seems that it doesn't affact the quarkus-keycloak-authorization.

**Reply by SetoKaiba** [reply]
![image](https://user-images.githubusercontent.com/159540/217297738-6f7c14c7-cd31-4a01-a32f-d6f90f6ded5e.png)
The enforcementMode here is still enforcing.

**Reply by SetoKaiba** [reply]
I thought that "the quarkus-keycloak-authorization and policy enforcer enabled" means the authorization is all managed by keycloak authorization tab.

**Reply by SetoKaiba** [reply]
This setting solved my problem. I originally thought that the enforcement-mode is come from keycloak.
```
# Enable Policy Enforcement
quarkus.keycloak.policy-enforcer.enable=true
quarkus.keycloak.policy-enforcer.enforcement-mode=permissive
```

**Reply by SetoKaiba** [reply]
![image](https://user-images.githubusercontent.com/159540/217299697-c2dd6502-7092-4730-9540-9ede323b4397.png)
So does it mean that this settings is not working with quarkus-keycloak-authorization? The enforcement mode must be set by application.properties?

**Reply by SetoKaiba** [reply]
And also, a final question. How can I do a task for more than 2000ms with request scoped SecurityIdentity, UserInfo or JsonWebToken?
If I make a sub thread, the request scoped stuffs are reporting no active request scope problem.
If I make in main thread, onOpen will tell me that it costs more than 2000ms, and some work is lost.

**Reply by sberyozkin** [reply]
@SetoKaiba great work, thanks for finding the solution which works.

**Reply by SetoKaiba** [reply]
Thank you for your help. What about the final 2 questions? Or should I create a new discussion thread?
- The enforcement mode I think that it should the same with the client settings if the enforcement mode is not set explicitly in application.properties. Do you think that it's a bug?
- How can I use SecurityIdentity, UserInfo and JsonWebToken in a sub thread? They're all request scoped beans. This problem is caused by the onOpen method can't take time longer than 2000ms.

**Comment by sberyozkin** [comment]
> The enforcement mode I think that it should the same with the client settings if the enforcement mode is not set explicitly in application.properties. Do you think that it's a bug?

Can you clarify please with the example (CC @pedroigor )

>How can I use SecurityIdentity, UserInfo and JsonWebToken in a sub thread? They're all request scoped beans.

Hi @manovotn @mkouba, I recall we were discussing awhile back how to correctly handle the request scope propagation through the diff thread contexts, is it about having `smallrye-context-propagation` and `@ActivateRequestContext` ? What are the current recommendations ? thanks

**Reply by SetoKaiba** [reply]
> Can you clarify please with the example (CC @pedroigor )

![image](https://user-images.githubusercontent.com/159540/217332581-a0913c14-6278-4278-8d33-d73cd7dbeba4.png)
![image](https://user-images.githubusercontent.com/159540/217333486-54430239-ab35-4c1b-86e9-527768581092.png)

As my previous post said, I set this to permissive. And according to the debug you instruct about the apply method. I found that the AbstractPolicyEnfocer.authorize method use the Enforcing mode instead of the Permissive settings in the Keycloak -> Clients -> your client -> Authorization -> Settings -> Policy enforcement mode.

**Reply by SetoKaiba** [reply]
> Hi @manovotn @mkouba, I recall we were discussing awhile back how to correctly handle the request scope propagation through the diff thread contexts, is it about having `smallrye-context-propagation` and `@ActivateRequestContext` ? What are the current recommendations ? thanks

https://quarkus.io/guides/security-customization#reactive-security
Just check the post below. It seems that it's a bug that WS doesn't support context propagation.

**Reply by SetoKaiba** [reply]
It seems that I can just use ManagedExecutor directly. The context propagation is working correctly for REST.
But it seems that the context propagation is not working for websocket.

![image](https://user-images.githubusercontent.com/159540/217369205-aa1ec929-5edc-4cc7-b463-33edeb2d2387.png)
![image](https://user-images.githubusercontent.com/159540/217369829-33bb7f8f-41d6-49a7-9bfd-3e2203913aeb.png)
For REST, they all run with with executor-threads. It does propagate the context.
![image](https://user-images.githubusercontent.com/159540/217369889-58a82397-c228-45d9-a60f-7659f6bd8a81.png)
![image](https://user-images.githubusercontent.com/159540/217369948-d7cf1663-00ef-47e3-9ecc-9a0f7ae14914.png)
For WS, one run with vert.x-eventloop-thread and one run with executor-thread. It doesn't propogate the context.

**Comment by sberyozkin** [comment]
@SetoKaiba 

>It seems that I can just use ManagedExecutor directly. The context propagation is working correctly for REST.
But it seems that the context propagation is not working for websocket.

Makes sense to create a bug (or enhancement) request as I'm not sure the context propagation was meant to support WS calls.
Please try to create a simple reproducer, without the security involved, to simplify the setup

**Reply by SetoKaiba** [reply]
I'd like to create a bug. But I have no idea that what other request scoped bean can be injected to validate the context propagation failure without security. Maybe I can create a issue first refer this post? And others can provide some effort to validate the context propagation failure.

**Reply by SetoKaiba** [reply]
https://github.com/quarkusio/quarkus/issues/30994
Here's the link for the bug.

**Reply by SetoKaiba** [reply]
@sberyozkin It seems that no one answers me about the context propagation bug.
